File: final/src/gclient.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <pthread.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <signal.h>
#include <time.h>
#include <sys/select.h>
#include <errno.h>

#define BUFFER_SIZE 2048
#define MAX_MESSAGE_LEN 1024

// ANSI color codes
#define COLOR_RED     "\033[91m"
#define COLOR_GREEN   "\033[92m"
#define COLOR_YELLOW  "\033[93m"
#define COLOR_BLUE    "\033[94m"
#define COLOR_PURPLE  "\033[95m"
#define COLOR_CYAN    "\033[96m"
#define COLOR_WHITE   "\033[97m"
#define COLOR_RESET   "\033[0m"
#define COLOR_DEBUG   "\033[90m"

#define DEBUG 1

#if DEBUG
#define debug_print(fmt, ...) printf(COLOR_DEBUG fmt COLOR_RESET, ##__VA_ARGS__)
#else
#define debug_print(fmt, ...)
#endif

typedef struct {
    int socket;
    int running;
    pthread_t receiver_thread;
} client_t;

client_t client = {0};

// Function prototypes
void signal_handler(int sig);
void print_colored(const char *text, const char *color);
void show_help(void);
char* extract_json_value(const char *json, const char *key);
void process_message(const char *message_str);
void* receive_messages(void *arg);
int connect_to_server(const char *host, int port);
void handle_input(void);
void disconnect_from_server(void);

// Signal handler for graceful shutdown
void signal_handler(int sig) {
    if (sig == SIGINT) {
        print_colored("Disconnecting...", COLOR_YELLOW);
        debug_print("[DEBUG] SIGINT received, shutting down client...\n");
        client.running = 0;
        disconnect_from_server();
        exit(0); // Ensure the program terminates
    }
}

// Print text with color
void print_colored(const char *text, const char *color) {
    printf("%s%s%s\n", color, text, COLOR_RESET);
}

// Show help message
void show_help(void) {
    print_colored("\nAvailable Commands:", COLOR_CYAN);
    print_colored("/join <room_name>     - Join or create a room", COLOR_CYAN);
    print_colored("/leave                - Leave current room", COLOR_CYAN);
    print_colored("/broadcast <message>  - Send message to everyone in current room", COLOR_CYAN);
    print_colored("/whisper <user> <msg> - Send private message to user", COLOR_CYAN);
    print_colored("/exit                 - Disconnect from server", COLOR_CYAN);
    print_colored("/help                 - Show this help message", COLOR_CYAN);
}

void process_message(const char *message_str) {
    debug_print("[DEBUG] Raw message received: %s\n", message_str);

    // Buffers to hold values safely
    char type[64] = {0}, content[512] = {0}, status[64] = {0}, timestamp[64] = {0};

    // Look for each key and extract its value manually
    const char *type_key = "\"type\":\"";
    const char *content_key = "\"content\":\"";
    const char *status_key = "\"status\":\"";
    const char *timestamp_key = "\"timestamp\":\"";

    const char *start, *end;

    // Extract "type"
    start = strstr(message_str, type_key);
    if (start) {
        start += strlen(type_key);
        end = strchr(start, '"');
        if (end && (end - start) < sizeof(type))
            strncpy(type, start, end - start);
    }

    // Extract "content"
    start = strstr(message_str, content_key);
    if (start) {
        start += strlen(content_key);
        end = strchr(start, '"');
        if (end && (end - start) < sizeof(content))
            strncpy(content, start, end - start);
    }

    // Extract "status"
    start = strstr(message_str, status_key);
    if (start) {
        start += strlen(status_key);
        end = strchr(start, '"');
        if (end && (end - start) < sizeof(status))
            strncpy(status, start, end - start);
    }

    // Extract "timestamp"
    start = strstr(message_str, timestamp_key);
    if (start) {
        start += strlen(timestamp_key);
        end = strchr(start, '"');
        if (end && (end - start) < sizeof(timestamp))
            strncpy(timestamp, start, end - start);
    }

    // Ensure null-termination
    type[sizeof(type)-1] = '\0';
    content[sizeof(content)-1] = '\0';
    status[sizeof(status)-1] = '\0';
    timestamp[sizeof(timestamp)-1] = '\0';

    if (type[0] == '\0' || content[0] == '\0') {
        print_colored("[ERROR] Malformed message received", COLOR_RED);
        debug_print("[DEBUG] Malformed message: %s\n", message_str);
        return;
    }

    debug_print("[DEBUG] Extracted type: %s, content: %s, status: %s, timestamp: %s\n",
           type, content, status[0] ? status : "N/A", timestamp[0] ? timestamp : "N/A");

    char formatted_message[BUFFER_SIZE];

    if (strcmp(type, "system") == 0) {
        snprintf(formatted_message, sizeof(formatted_message), "[SYSTEM] %s", content);
        print_colored(formatted_message, COLOR_CYAN);
    } else if (strcmp(type, "success") == 0) {
        snprintf(formatted_message, sizeof(formatted_message), "[SUCCESS] %s", content);
        print_colored(formatted_message, COLOR_GREEN);
    } else if (strcmp(type, "error") == 0) {
        snprintf(formatted_message, sizeof(formatted_message), "[ERROR] %s", content);
        print_colored(formatted_message, COLOR_RED);
    } else if (strcmp(type, "broadcast") == 0) {
        snprintf(formatted_message, sizeof(formatted_message), "[%s] %s",
                 timestamp[0] ? timestamp : "", content);
        print_colored(formatted_message, COLOR_YELLOW);
    } else if (strcmp(type, "whisper") == 0) {
        snprintf(formatted_message, sizeof(formatted_message), "[Private] [%s] %s",
                 timestamp[0] ? timestamp : "", content);
        print_colored(formatted_message, COLOR_PURPLE);
    } else if (strcmp(type, "notification") == 0) {
        snprintf(formatted_message, sizeof(formatted_message), "[INFO] %s", content);
        print_colored(formatted_message, COLOR_BLUE);
    } else {
        snprintf(formatted_message, sizeof(formatted_message), "[UNKNOWN] [%s] %s",
                 timestamp[0] ? timestamp : "", content);
        print_colored(formatted_message, COLOR_WHITE);
    }

    debug_print("[DEBUG] Processed message type: %s, content: %s\n", type, content);
}



// Receive messages from server thread
void* receive_messages(void *arg) {
    char buffer[BUFFER_SIZE];
    char message_buffer[BUFFER_SIZE * 2] = {0};

    fd_set read_fds;
    struct timeval timeout;

    while (client.running) {
        FD_ZERO(&read_fds);
        FD_SET(client.socket, &read_fds);

        timeout.tv_sec = 1;  // 1 second timeout
        timeout.tv_usec = 0;

        int activity = select(client.socket + 1, &read_fds, NULL, NULL, &timeout);

        if (activity < 0 && errno != EINTR) {
            perror("select error");
            break;
        }

        if (activity == 0) {
            // Timeout, no data, check if we should still run
            continue;
        }

        if (FD_ISSET(client.socket, &read_fds)) {
            int bytes_received = recv(client.socket, buffer, sizeof(buffer) - 1, 0);
            if (bytes_received <= 0) {
                if (client.running) {
                    print_colored("Connection lost or closed by server", COLOR_RED);
                }
                break;
            }

            buffer[bytes_received] = '\0';
            strcat(message_buffer, buffer);

            // Process complete messages (ending with newline)
            char *newline_pos;
            while ((newline_pos = strchr(message_buffer, '\n')) != NULL) {
                *newline_pos = '\0';
                if (strlen(message_buffer) > 0) {
                    process_message(message_buffer);
                }
                // Move remaining data to beginning of buffer
                memmove(message_buffer, newline_pos + 1, strlen(newline_pos + 1) + 1);
            }
        }
    }

    return NULL;
}

// Connect to server
int connect_to_server(const char *host, int port) {
    struct sockaddr_in server_addr;
    
    // Create socket
    client.socket = socket(AF_INET, SOCK_STREAM, 0);
    if (client.socket < 0) {
        perror("Socket creation failed");
        return 0;
    }
    
    // Configure server address
    memset(&server_addr, 0, sizeof(server_addr));
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(port);
    
    if (inet_pton(AF_INET, host, &server_addr.sin_addr) <= 0) {
        perror("Invalid address");
        close(client.socket);
        return 0;
    }
    
    // Connect to server
    if (connect(client.socket, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
        perror("Connection failed");
        close(client.socket);
        return 0;
    }
    
    client.running = 1;
    
    char success_msg[256];
    snprintf(success_msg, sizeof(success_msg), "Connected to server at %s:%d", host, port);
    print_colored(success_msg, COLOR_GREEN);
    
    // Start receiver thread
    if (pthread_create(&client.receiver_thread, NULL, receive_messages, NULL) != 0) {
        perror("Thread creation failed");
        close(client.socket);
        client.running = 0;
        return 0;
    }
    
    return 1;
}

// Handle user input
void handle_input(void) {
    char input[MAX_MESSAGE_LEN];
    
    while (client.running) {
        if (fgets(input, sizeof(input), stdin) == NULL) {
            break;
        }
        
        // Remove newline
        char *newline = strchr(input, '\n');
        if (newline) *newline = '\0';
        
        // Skip empty input
        if (strlen(input) == 0) {
            continue;
        }
        
        // Handle local commands
        if (strcmp(input, "/help") == 0) {
            show_help();
            continue;
        }
        
        if (strcmp(input, "/exit") == 0) {
            client.running = 0;
            break;
        }
        
        // Send message to server
        if (send(client.socket, input, strlen(input), 0) < 0) {
            print_colored("Failed to send message", COLOR_RED);
            break;
        }
    }
}

// Disconnect from server
void disconnect_from_server(void) {
    client.running = 0;
    
    if (client.socket > 0) {
        close(client.socket);
        client.socket = 0;
    }
    
    // Wait for receiver thread to finish
    if (client.receiver_thread) {
        pthread_join(client.receiver_thread, NULL);
    }
    
    print_colored("Disconnected from server", COLOR_YELLOW);
}

int main(int argc, char *argv[]) {
    char *host = "127.0.0.1";
    int port = 8888;
    
    // Parse command line arguments
    if (argc >= 3) {
        host = argv[1];
        port = atoi(argv[2]);
    } else if (argc >= 2) {
        port = atoi(argv[1]);
    }
    
    if (port <= 0 || port > 65535) {
        fprintf(stderr, "Invalid port number\n");
        exit(EXIT_FAILURE);
    }
    
    // Setup signal handler
    signal(SIGINT, signal_handler);
    
    printf("=== Multi-threaded Chat Client ===\n");
    printf("Connecting to %s:%d...\n", host, port);
    printf("Type '/help' for available commands\n");
    printf("Press Ctrl+C to quit\n\n");
    
    if (!connect_to_server(host, port)) {
        fprintf(stderr, "Failed to connect to server\n");
        exit(EXIT_FAILURE);
    }
    
    // Handle user input
    handle_input();
    
    // Clean up
    disconnect_from_server();
    
    printf("Connection closed\n");
    return 0;
}

File: final/client/gclient.py
#!/usr/bin/env python3
"""
Exact Python replica of the C chat client - Compatible with C server
"""

import socket
import threading
import time
import signal
import sys
import os
import json
import select
from datetime import datetime

# Constants matching C client exactly
BUFFER_SIZE = 2048
MAX_MESSAGE_LEN = 1024
MAX_FILE_SIZE = 3 * 1024 * 1024  # 3MB

# ANSI color codes - exact match to C client
COLOR_RED = "\033[91m"
COLOR_GREEN = "\033[92m"
COLOR_YELLOW = "\033[93m"
COLOR_BLUE = "\033[94m"
COLOR_PURPLE = "\033[95m"
COLOR_CYAN = "\033[96m"
COLOR_WHITE = "\033[97m"
COLOR_RESET = "\033[0m"
COLOR_DEBUG = "\033[90m"

DEBUG = False

def debug_print(message):
    """Debug print function matching C client"""
    if DEBUG:
        print(f"{COLOR_DEBUG}{message}{COLOR_RESET}")

def print_colored(text, color):
    """Print text with color - exact C client match"""
    print(f"{color}{text}{COLOR_RESET}")
    sys.stdout.flush()

def show_help():
    """Show help message - exact C client match"""
    print_colored("\nAvailable Commands:", COLOR_CYAN)
    print_colored("/join <room_name>     - Join or create a room", COLOR_CYAN)
    print_colored("/leave                - Leave current room", COLOR_CYAN)
    print_colored("/broadcast <message>  - Send message to everyone in current room", COLOR_CYAN)
    print_colored("/whisper <user> <msg> - Send private message to user", COLOR_CYAN)
    print_colored("/sendfile <file> <user> - Send file to user (max 3MB, .txt/.pdf/.jpg/.png)", COLOR_CYAN)
    print_colored("/exit                 - Disconnect from server", COLOR_CYAN)
    print_colored("/help                 - Show this help message", COLOR_CYAN)

def validate_file(filename):
    """Validate file for sending - exact C client logic"""
    if not filename or len(filename) == 0:
        return False
    
    # Check if file exists and get size
    try:
        stat_info = os.stat(filename)
    except OSError:
        return False  # File doesn't exist
    
    # Check file size (max 3MB)
    if stat_info.st_size > MAX_FILE_SIZE:
        return False
    
    # Check file extension
    ext = os.path.splitext(filename)[1].lower()
    if ext in ['.txt', '.pdf', '.jpg', '.png']:
        return True
    
    return False

def get_file_size(filename):
    """Get file size in bytes - exact C client logic"""
    try:
        return os.path.getsize(filename)
    except OSError:
        return -1

class ChatClient:
    def __init__(self):
        self.socket = None
        self.running = False
        self.receiver_thread = None
        
        # Setup signal handler
        signal.signal(signal.SIGINT, self.signal_handler)

    def signal_handler(self, signum, frame):
        """Handle SIGINT for graceful shutdown - exact C client match"""
        print_colored("Disconnecting...", COLOR_YELLOW)
        debug_print("[DEBUG] SIGINT received, shutting down client...")
        self.running = False
        self.disconnect_from_server()
        sys.exit(0)

    def extract_json_fields(self, message_str):
        """Extract JSON fields manually for better compatibility with C version"""
        type_val = content_val = status_val = timestamp_val = ""
        
        # Extract type
        type_start = message_str.find('"type":"')
        if type_start != -1:
            type_start += 8
            type_end = message_str.find('"', type_start)
            if type_end != -1:
                type_val = message_str[type_start:type_end]
        
        # Extract content
        content_start = message_str.find('"content":"')
        if content_start != -1:
            content_start += 11
            content_end = message_str.find('"', content_start)
            if content_end != -1:
                content_val = message_str[content_start:content_end]
        
        # Extract status
        status_start = message_str.find('"status":"')
        if status_start != -1:
            status_start += 10
            status_end = message_str.find('"', status_start)
            if status_end != -1:
                status_val = message_str[status_start:status_end]
        
        # Extract timestamp
        timestamp_start = message_str.find('"timestamp":"')
        if timestamp_start != -1:
            timestamp_start += 13
            timestamp_end = message_str.find('"', timestamp_start)
            if timestamp_end != -1:
                timestamp_val = message_str[timestamp_start:timestamp_end]
        
        return type_val, content_val, status_val, timestamp_val

    def process_message(self, message_str):
        """Process incoming message - exact C client logic"""
        debug_print(f"[DEBUG] Raw message received: {message_str}")
        
        # Extract JSON fields manually like C client
        msg_type, content, status, timestamp = self.extract_json_fields(message_str)
        
        if not msg_type or not content:
            print_colored("[ERROR] Malformed message received", COLOR_RED)
            debug_print(f"[DEBUG] Malformed message: {message_str}")
            return
        
        debug_print(f"[DEBUG] Extracted type: {msg_type}, content: {content}, status: {status or 'N/A'}, timestamp: {timestamp or 'N/A'}")
        
        if msg_type == "system":
            formatted_message = f"[SYSTEM] {content}"
            print_colored(formatted_message, COLOR_CYAN)
            
        elif msg_type == "success":
            formatted_message = f"[SUCCESS] {content}"
            print_colored(formatted_message, COLOR_GREEN)
            
        elif msg_type == "error":
            formatted_message = f"[ERROR] {content}"
            print_colored(formatted_message, COLOR_RED)
            
        elif msg_type == "broadcast":
            formatted_message = f"[{timestamp}] {content}" if timestamp else content
            print_colored(formatted_message, COLOR_YELLOW)
            
        elif msg_type == "whisper":
            formatted_message = f"[Private] [{timestamp}] {content}" if timestamp else f"[Private] {content}"
            print_colored(formatted_message, COLOR_PURPLE)
            
        elif msg_type == "notification":
            formatted_message = f"[INFO] {content}"
            print_colored(formatted_message, COLOR_BLUE)
            
        elif msg_type == "file_request":
            formatted_message = f"[FILE REQUEST] {content}"
            print_colored(formatted_message, COLOR_CYAN)
            
        elif msg_type == "file_incoming":
            formatted_message = f"[FILE INCOMING] {content}"
            print_colored(formatted_message, COLOR_CYAN)
            print_colored("[INFO] File is being prepared by server...", COLOR_CYAN)
            
        elif msg_type == "file_transfer_start":
            # Handle file transfer start - exact C client logic
            self.handle_file_transfer_start(content)
            
        elif msg_type == "file_transfer_end":
            # Handle file transfer completion notification
            self.handle_file_transfer_end(content)
            
        elif msg_type == "file_complete":
            formatted_message = f"[SUCCESS] [{timestamp}] {content}" if timestamp else f"[SUCCESS] {content}"
            print_colored(formatted_message, COLOR_GREEN)
            
        elif msg_type == "filename_proposal":
            # Handle filename proposal from server - exact C client logic
            self.handle_filename_proposal(content)
            
        else:
            formatted_message = f"[UNKNOWN] [{timestamp}] {content}" if timestamp else f"[UNKNOWN] {content}"
            print_colored(formatted_message, COLOR_WHITE)
        
        debug_print(f"[DEBUG] Processed message type: {msg_type}, content: {content}")

    def handle_file_transfer_start(self, content):
        """Handle file transfer start - exact C client logic"""
        # Parse content: "FILE_TRANSFER_START:filename:sender:size"
        parts = content.split(':')
        if len(parts) >= 4 and parts[0] == "FILE_TRANSFER_START":
            filename = parts[1]
            sender = parts[2]
            try:
                file_size = int(parts[3])
            except ValueError:
                print_colored("[ERROR] Invalid file transfer start message", COLOR_RED)
                return
            
            formatted_message = f"[FILE INCOMING] Receiving '{filename}' from {sender} ({file_size / 1024:.1f} KB)..."
            print_colored(formatted_message, COLOR_CYAN)
            
            # Prepare to receive file data
            try:
                with open(filename, "wb") as file:
                    bytes_received = 0
                    
                    print_colored("[INFO] Receiving file data...", COLOR_CYAN)
                    
                    # Receive file data in chunks
                    while bytes_received < file_size:
                        bytes_to_read = min(8192, file_size - bytes_received)
                        
                        try:
                            data = self.socket.recv(bytes_to_read)
                            if not data:
                                print_colored("[ERROR] Connection lost during file transfer", COLOR_RED)
                                os.unlink(filename)  # Remove incomplete file
                                return
                            
                            file.write(data)
                            bytes_received += len(data)
                            
                        except socket.error:
                            print_colored("[ERROR] Failed to receive file data", COLOR_RED)
                            os.unlink(filename)  # Remove incomplete file
                            return
                    
                    if bytes_received == file_size:
                        formatted_message = f"[FILE SUCCESS] File '{filename}' received from {sender} ({bytes_received / 1024:.1f} KB) - saved locally"
                        print_colored(formatted_message, COLOR_GREEN)
                    else:
                        print_colored("[ERROR] File transfer incomplete", COLOR_RED)
                        os.unlink(filename)  # Remove incomplete file
                        
            except IOError:
                print_colored("[ERROR] Cannot create file for writing", COLOR_RED)
        else:
            print_colored("[ERROR] Invalid file transfer start message", COLOR_RED)

    def handle_file_transfer_end(self, content):
        """Handle file transfer completion notification - exact C client logic"""
        # Parse content: "FILE_TRANSFER_END:filename:size"
        parts = content.split(':')
        if len(parts) >= 3 and parts[0] == "FILE_TRANSFER_END":
            filename = parts[1]
            try:
                file_size = int(parts[2])
                formatted_message = f"[FILE COMPLETE] Transfer of '{filename}' completed ({file_size / 1024:.1f} KB)"
                print_colored(formatted_message, COLOR_GREEN)
            except ValueError:
                print_colored("[ERROR] Invalid file transfer end message", COLOR_RED)

    def handle_filename_proposal(self, content):
        """Handle filename proposal from server - exact C client logic"""
        # Parse content: "FILENAME_PROPOSAL:filename:sender:size"
        parts = content.split(':')
        if len(parts) >= 4 and parts[0] == "FILENAME_PROPOSAL":
            proposed_filename = parts[1]
            sender = parts[2]
            try:
                file_size = int(parts[3])
            except ValueError:
                print_colored("[ERROR] Invalid filename proposal message", COLOR_RED)
                self.socket.send(b"FILENAME_OK")  # Send OK as fallback
                return
            
            formatted_message = f"[FILE PROPOSAL] {sender} wants to send '{proposed_filename}' ({file_size / 1024:.1f} KB)"
            print_colored(formatted_message, COLOR_CYAN)
            sys.stdout.flush()
            
            # Automatically check if file already exists locally
            file_exists = os.path.exists(proposed_filename)
            
            if file_exists:
                try:
                    existing_size = os.path.getsize(proposed_filename)
                    conflict_msg = f"[WARNING] File '{proposed_filename}' already exists locally ({existing_size / 1024:.1f} KB) - requesting rename"
                    print_colored(conflict_msg, COLOR_YELLOW)
                    sys.stdout.flush()
                    
                    # Automatically request filename change to avoid overwriting
                    self.socket.send(b"FILENAME_NOT")
                    print_colored("[INFO] Filename will be automatically renamed by server", COLOR_CYAN)
                    sys.stdout.flush()
                except OSError:
                    # File exists but can't get size, still request rename
                    self.socket.send(b"FILENAME_NOT")
                    print_colored("[INFO] Filename will be automatically renamed by server", COLOR_CYAN)
                    sys.stdout.flush()
            else:
                # File doesn't exist, accept original filename
                self.socket.send(b"FILENAME_OK")
                print_colored("[INFO] Filename accepted. Preparing to receive file...", COLOR_GREEN)
                sys.stdout.flush()
        else:
            print_colored("[ERROR] Invalid filename proposal message", COLOR_RED)
            sys.stdout.flush()
            # Send OK as fallback
            self.socket.send(b"FILENAME_OK")

    def receive_messages(self):
        """Receive messages from server thread - exact C client logic"""
        message_buffer = ""
        
        while self.running:
            # Use select with timeout like C client
            ready, _, _ = select.select([self.socket], [], [], 1.0)
            
            if not ready:
                continue  # Timeout, check if we should still run
            
            try:
                data = self.socket.recv(BUFFER_SIZE)
                if not data:
                    if self.running:
                        print_colored("Connection lost or closed by server", COLOR_RED)
                    break
                
                message_buffer += data.decode('utf-8')
                
                # Process complete messages (ending with newline)
                while '\n' in message_buffer:
                    message, message_buffer = message_buffer.split('\n', 1)
                    if message.strip():
                        self.process_message(message.strip())
                        
            except socket.error as e:
                if self.running:
                    print_colored(f"Socket error: {e}", COLOR_RED)
                break
            except UnicodeDecodeError:
                # Skip invalid UTF-8 data
                continue

    def connect_to_server(self, host, port):
        """Connect to server - exact C client logic"""
        try:
            # Create socket
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            
            # Connect to server
            self.socket.connect((host, port))
            
            self.running = True
            
            success_msg = f"Connected to server at {host}:{port}"
            print_colored(success_msg, COLOR_GREEN)
            
            # Start receiver thread
            self.receiver_thread = threading.Thread(target=self.receive_messages, daemon=True)
            self.receiver_thread.start()
            
            return True
            
        except socket.error as e:
            print(f"Connection failed: {e}")
            if self.socket:
                self.socket.close()
            return False

    def handle_sendfile_command(self, input_line):
        """Handle /sendfile command locally - exact C client logic"""
        parts = input_line.split(' ')
        if len(parts) < 3:
            print_colored("[ERROR] Usage: /sendfile <filename> <username>", COLOR_RED)
            return
        
        filename = parts[1]
        username = parts[2]
        
        # Validate file exists and get size
        file_size = get_file_size(filename)
        if file_size < 0:
            print_colored("[ERROR] File not found or cannot be accessed", COLOR_RED)
            return
        
        # Validate file type and size
        if not validate_file(filename):
            print_colored("[ERROR] Invalid file type or size. Allowed: .txt, .pdf, .jpg, .png (max 3MB)", COLOR_RED)
            return
        
        confirm_msg = f"[INFO] Sending file '{filename}' ({file_size / 1024:.1f} KB) to {username}..."
        print_colored(confirm_msg, COLOR_CYAN)
        
        # Send command with file size to server
        sendfile_cmd = f"/sendfile {filename} {username} {file_size}"
        
        try:
            self.socket.send(sendfile_cmd.encode('utf-8'))
            
            # Wait for server response - either error or ready for file data
            response = self.socket.recv(BUFFER_SIZE).decode('utf-8')
            
            # Check if server is ready for file data or gave an error
            if "Ready to receive file data" in response:
                print_colored("[INFO] Uploading file data...", COLOR_CYAN)
                
                # Send file data
                try:
                    with open(filename, "rb") as file:
                        bytes_sent = 0
                        
                        # Send file data in chunks
                        while True:
                            chunk = file.read(8192)
                            if not chunk:
                                break
                            
                            self.socket.send(chunk)
                            bytes_sent += len(chunk)
                        
                        if bytes_sent == file_size:
                            print_colored("[INFO] File upload completed. Processing...", COLOR_GREEN)
                        else:
                            print_colored("[ERROR] File upload incomplete", COLOR_RED)
                            
                except IOError:
                    print_colored("[ERROR] Failed to open file for reading", COLOR_RED)
            else:
                # Server gave an error, process the response
                self.process_message(response)
                
        except socket.error:
            print_colored("[ERROR] Failed to send command to server", COLOR_RED)

    def handle_input(self):
        """Handle user input - exact C client logic"""
        while self.running:
            try:
                user_input = input().strip()
                
                # Skip empty input
                if not user_input:
                    continue
                
                # Handle local commands
                if user_input == "/help":
                    show_help()
                    continue
                
                if user_input == "/exit":
                    self.running = False
                    break
                
                # Handle /sendfile command locally
                if user_input.startswith("/sendfile "):
                    self.handle_sendfile_command(user_input)
                    continue
                
                # Send message to server
                self.socket.send(user_input.encode('utf-8'))
                
            except EOFError:
                break
            except socket.error:
                print_colored("Failed to send message", COLOR_RED)
                break

    def disconnect_from_server(self):
        """Disconnect from server - exact C client logic"""
        self.running = False
        
        if self.socket:
            self.socket.close()
            self.socket = None
        
        # Wait for receiver thread to finish
        if self.receiver_thread and self.receiver_thread.is_alive():
            self.receiver_thread.join(timeout=2.0)
        
        print_colored("Disconnected from server", COLOR_YELLOW)

def main():
    host = "127.0.0.1"
    port = 8888
    
    # Parse command line arguments - exact C client logic
    if len(sys.argv) >= 3:
        host = sys.argv[1]
        try:
            port = int(sys.argv[2])
        except ValueError:
            print("Invalid port number", file=sys.stderr)
            sys.exit(1)
    elif len(sys.argv) >= 2:
        try:
            port = int(sys.argv[1])
        except ValueError:
            print("Invalid port number", file=sys.stderr)
            sys.exit(1)
    
    if port <= 0 or port > 65535:
        print("Invalid port number", file=sys.stderr)
        sys.exit(1)
    
    print("=== Multi-threaded Chat Client ===")
    print(f"Connecting to {host}:{port}...")
    print("Type '/help' for available commands")
    print("Press Ctrl+C to quit\n")
    
    client = ChatClient()
    
    if not client.connect_to_server(host, port):
        print("Failed to connect to server", file=sys.stderr)
        sys.exit(1)
    
    # Handle user input
    client.handle_input()
    
    # Clean up
    client.disconnect_from_server()
    
    print("Connection closed")

if __name__ == "__main__":
    main()